File: server/routes/posts.ts
/**
 * The community forum: posts, votes and comments.
 *
 * Nothing here is linked at the schema level — votes and comments carry a
 * `postId` with no foreign key behind it — so each handler that writes one
 * checks the post exists first, and deleting a post takes its votes and
 * comments with it. An orphaned row is invisible and permanent.
 */

import { Hono } from "hono";
import { getUser } from "../kinde";
import { db } from "../db";
import { posts, postComments, postVotes } from "../db/schema/posts";
import { eq, desc, sql, and } from "drizzle-orm";
import { z } from "zod";
import { zValidator } from "@hono/zod-validator";
import { requireDb } from "../middleware";
import { authorName as displayName, idParam } from "../http";

const createPostSchema = z.object({
  title: z.string().min(1).max(200),
  body: z.string().max(5000).default(""),
  lessonPlanId: z.number().optional(),
  lessonPlanName: z.string().optional(),
});

const createCommentSchema = z.object({
  body: z.string().min(1).max(2000),
});

/** A user's standing vote on a post: up, down, or none. */
type VoteValue = 1 | -1 | 0;

/** 1 when this vote counts toward `side`'s tally, else 0. */
const tally = (value: VoteValue, side: 1 | -1): number => (value === side ? 1 : 0);

export const postsRoute = new Hono()
  .use("*", requireDb)

  // ── GET all posts (with sort) ──
  .get("/", async (c) => {
    const sort = c.req.query("sort") || "latest";
    const query = db!.select().from(posts);
    const result = sort === "top"
      ? await query.orderBy(desc(posts.upvotes))
      : await query.orderBy(desc(posts.createdAt));
    return c.json({ posts: result });
  })

  // ── CREATE post ──
  .post("/", getUser, zValidator("json", createPostSchema), async (c) => {
    const user = c.var.user;
    const data = c.req.valid("json");
    const authorName = displayName(user);
    const [post] = await db!.insert(posts).values({
      userId: user.id,
      authorName,
      title: data.title,
      body: data.body,
      lessonPlanId: data.lessonPlanId ?? null,
      lessonPlanName: data.lessonPlanName ?? null,
    }).returning();
    return c.json({ post }, 201);
  })

  // ── GET single post ──
  .get("/:id", async (c) => {
    const id = idParam(c.req.param("id"));
    if (id === null) return c.json({ error: "Invalid post id" }, 400);
    const [post] = await db!.select().from(posts).where(eq(posts.id, id));
    if (!post) return c.json({ error: "Not found" }, 404);
    const comments = await db!.select().from(postComments)
      .where(eq(postComments.postId, id))
      .orderBy(desc(postComments.createdAt));
    return c.json({ post, comments });
  })

  // ── VOTE ──
  // Voting the same way twice clears the vote; voting the other way switches
  // it. Both counters are then adjusted by the difference between the old
  // state and the new one, in a single statement — this used to be a nested
  // if/else over six near-identical UPDATEs, one per transition, and the two
  // "switch" branches were the only ones that touched both columns.
  .post("/:id/vote", getUser, zValidator("json", z.object({ vote: z.union([z.literal(1), z.literal(-1)]) })), async (c) => {
    const user = c.var.user;
    const postId = idParam(c.req.param("id"));
    if (postId === null) return c.json({ error: "Invalid post id" }, 400);
    const { vote } = c.req.valid("json");

    const [target] = await db!.select({ id: posts.id }).from(posts).where(eq(posts.id, postId));
    if (!target) return c.json({ error: "Post not found" }, 404);

    const [existing] = await db!.select().from(postVotes)
      .where(and(eq(postVotes.postId, postId), eq(postVotes.userId, user.id)));

    const before: VoteValue = existing ? (existing.vote as 1 | -1) : 0;
    const after: VoteValue = before === vote ? 0 : vote;

    if (after === 0) {
      await db!.delete(postVotes).where(eq(postVotes.id, existing.id));
    } else if (existing) {
      await db!.update(postVotes).set({ vote: after }).where(eq(postVotes.id, existing.id));
    } else {
      await db!.insert(postVotes).values({ postId, userId: user.id, vote: after });
    }

    const [updated] = await db!.update(posts)
      .set({
        // GREATEST floors at zero: a count that drifted negative — which a
        // half-applied vote in the old branching version could do — renders
        // as "-1 upvotes".
        upvotes: sql`GREATEST(${posts.upvotes} + ${tally(after, 1) - tally(before, 1)}, 0)`,
        downvotes: sql`GREATEST(${posts.downvotes} + ${tally(after, -1) - tally(before, -1)}, 0)`,
      })
      .where(eq(posts.id, postId))
      .returning();

    return c.json({ post: updated });
  })

  // ── ADD COMMENT ──
  .post("/:id/comments", getUser, zValidator("json", createCommentSchema), async (c) => {
    const user = c.var.user;
    const postId = idParam(c.req.param("id"));
    if (postId === null) return c.json({ error: "Invalid post id" }, 400);
    const { body } = c.req.valid("json");
    const authorName = displayName(user);

    // Same reason as voting: an orphaned comment is invisible but permanent.
    const [target] = await db!.select({ id: posts.id }).from(posts).where(eq(posts.id, postId));
    if (!target) return c.json({ error: "Post not found" }, 404);

    const [comment] = await db!.insert(postComments).values({
      postId, userId: user.id, authorName, body,
    }).returning();

    await db!.update(posts).set({ commentCount: sql`${posts.commentCount} + 1` }).where(eq(posts.id, postId));
    return c.json({ comment }, 201);
  })

  // ── DELETE post ──
  // Only the author may delete. Votes and comments go too, so no rows are
  // orphaned against a post id that no longer exists.
  .delete("/:id", getUser, async (c) => {
    const user = c.var.user;
    const id = idParam(c.req.param("id"));
    if (id === null) return c.json({ error: "Invalid post id" }, 400);

    const [post] = await db!.select().from(posts).where(eq(posts.id, id));
    if (!post) return c.json({ error: "Post not found" }, 404);
    if (post.userId !== user.id) return c.json({ error: "You can only delete your own posts" }, 403);

    await db!.delete(postComments).where(eq(postComments.postId, id));
    await db!.delete(postVotes).where(eq(postVotes.postId, id));
    await db!.delete(posts).where(eq(posts.id, id));

    return c.json({ id });
  })

  // ── DELETE comment ──
  .delete("/:id/comments/:commentId", getUser, async (c) => {
    const user = c.var.user;
    const postId = idParam(c.req.param("id"));
    const commentId = idParam(c.req.param("commentId"));
    if (postId === null || commentId === null) {
      return c.json({ error: "Invalid id" }, 400);
    }

    const [comment] = await db!.select().from(postComments)
      .where(and(eq(postComments.id, commentId), eq(postComments.postId, postId)));
    if (!comment) return c.json({ error: "Comment not found" }, 404);
    if (comment.userId !== user.id) return c.json({ error: "You can only delete your own comments" }, 403);

    await db!.delete(postComments).where(eq(postComments.id, commentId));
    // Floor at zero: a count that drifted negative would render as "-1 comments".
    await db!.update(posts)
      .set({ commentCount: sql`GREATEST(${posts.commentCount} - 1, 0)` })
      .where(eq(posts.id, postId));

    return c.json({ id: commentId });
  });

File: backend/services/posts.py
"""The community forum: posts, votes and comments.

Nothing here is linked at the schema level — a vote carries a ``post_id`` with
no foreign key behind it — so each write checks the post exists first, and
deleting a post takes its votes and comments with it. An orphaned row is
invisible and permanent.
"""

from __future__ import annotations

from typing import Literal, Optional, Sequence

from sqlalchemy import delete, func, select, update
from sqlalchemy.ext.asyncio import AsyncSession

from ..core.errors import Forbidden, NotFound
from ..db.models import Post, PostComment, PostVote, User

SortMode = Literal["latest", "top"]

#: A standing vote: up, down, or none.
VoteValue = int


def _tally(value: VoteValue, side: int) -> int:
    """1 when this vote counts toward ``side``'s total, else 0."""
    return 1 if value == side else 0


async def list_posts(db: AsyncSession, sort: SortMode = "latest") -> Sequence[Post]:
    order = Post.upvotes.desc() if sort == "top" else Post.created_at.desc()
    return (await db.execute(select(Post).order_by(order))).scalars().all()


async def get_post(db: AsyncSession, post_id: int) -> Post:
    post = (
        await db.execute(select(Post).where(Post.id == post_id).limit(1))
    ).scalar_one_or_none()
    if post is None:
        raise NotFound("Post not found")
    return post


async def comments_for(db: AsyncSession, post_id: int) -> Sequence[PostComment]:
    return (
        (
            await db.execute(
                select(PostComment)
                .where(PostComment.post_id == post_id)
                .order_by(PostComment.created_at.desc())
            )
        )
        .scalars()
        .all()
    )


async def create_post(
    db: AsyncSession,
    author: User,
    *,
    title: str,
    body: str,
    lesson_plan_id: Optional[int],
    lesson_plan_name: Optional[str],
) -> Post:
    post = Post(
        user_id=author.id,
        author_name=display_name(author),
        title=title,
        body=body,
        lesson_plan_id=lesson_plan_id,
        lesson_plan_name=lesson_plan_name,
    )
    db.add(post)
    await db.flush()
    await db.refresh(post)
    return post


def display_name(user: User) -> str:
    """The name a byline shows: given name, else email handle, else "Member"."""
    if user.given_name and user.given_name.strip():
        return user.given_name.strip()
    if user.email and "@" in user.email:
        return user.email.split("@", 1)[0]
    return "Member"


async def cast_vote(db: AsyncSession, post_id: int, user_id: str, vote: int) -> Post:
    """Record a vote and return the post with its adjusted counts.

    Voting the same way twice clears the vote; voting the other way switches
    it. Both counters are then adjusted by the difference between the old state
    and the new one, in a single statement — this used to be a nested if/else
    over six near-identical UPDATEs, one per transition, and the two "switch"
    branches were the only ones that touched both columns.
    """
    await get_post(db, post_id)  # 404 before writing an orphaned vote

    existing = (
        await db.execute(
            select(PostVote).where(PostVote.post_id == post_id, PostVote.user_id == user_id).limit(1)
        )
    ).scalar_one_or_none()

    before: VoteValue = existing.vote if existing else 0
    after: VoteValue = 0 if before == vote else vote

    if after == 0 and existing is not None:
        await db.execute(delete(PostVote).where(PostVote.id == existing.id))
    elif existing is not None:
        existing.vote = after
    else:
        db.add(PostVote(post_id=post_id, user_id=user_id, vote=after))

    await db.execute(
        update(Post)
        .where(Post.id == post_id)
        .values(
            # GREATEST floors at zero: a count that drifted negative — which a
            # half-applied vote in the old branching version could do — renders
            # as "-1 upvotes".
            upvotes=func.greatest(Post.upvotes + (_tally(after, 1) - _tally(before, 1)), 0),
            downvotes=func.greatest(Post.downvotes + (_tally(after, -1) - _tally(before, -1)), 0),
        )
    )
    await db.flush()

    post = await get_post(db, post_id)
    await db.refresh(post)
    return post


async def add_comment(db: AsyncSession, post_id: int, author: User, body: str) -> PostComment:
    await get_post(db, post_id)  # same reason as voting

    comment = PostComment(
        post_id=post_id, user_id=author.id, author_name=display_name(author), body=body
    )
    db.add(comment)
    await db.execute(
        update(Post).where(Post.id == post_id).values(comment_count=Post.comment_count + 1)
    )
    await db.flush()
    await db.refresh(comment)
    return comment


async def delete_post(db: AsyncSession, post_id: int, user_id: str) -> None:
    """Only the author may delete. Votes and comments go too."""
    post = await get_post(db, post_id)
    if post.user_id != user_id:
        raise Forbidden("You can only delete your own posts")

    await db.execute(delete(PostComment).where(PostComment.post_id == post_id))
    await db.execute(delete(PostVote).where(PostVote.post_id == post_id))
    await db.execute(delete(Post).where(Post.id == post_id))


async def delete_comment(db: AsyncSession, post_id: int, comment_id: int, user_id: str) -> None:
    comment = (
        await db.execute(
            select(PostComment)
            .where(PostComment.id == comment_id, PostComment.post_id == post_id)
            .limit(1)
        )
    ).scalar_one_or_none()
    if comment is None:
        raise NotFound("Comment not found")
    if comment.user_id != user_id:
        raise Forbidden("You can only delete your own comments")

    await db.execute(delete(PostComment).where(PostComment.id == comment_id))
    # Floor at zero: a count that drifted negative would render as "-1 comments".
    await db.execute(
        update(Post)
        .where(Post.id == post_id)
        .values(comment_count=func.greatest(Post.comment_count - 1, 0))
    )
